### PR TITLE
Cascading 2.6 tracing

### PR DESCRIPTION
**Pull request by JamesIry**
When Scalding talked to document services like Driven it was sending line numbers that essentially mark the boundary between Scalding and Cascading. But that's not very useful to end users. A change was made to Cascading 2.6 that allows higher level APIs to indicate where a document service should consider the boundary. This PR bumps Scalding up to use Cascading 2.6 and to participate in the protocol for registering an API boundary.

### CODE REVIEW

**Comment by ianoc** [comment]
Thanks for the PR. Just an FYI we are going to be pretty slow about taking the major cascading bump as it'll take lots of testing.

File: scalding-core/src/main/scala/com/twitter/scalding/Job.scala [hunk 1]
```diff
@@ -77,6 +77,7 @@ object Job {
  * these functions can be combined Monadically using algebird.monad.Reader.
  */
 class Job(val args: Args) extends FieldConversions with java.io.Serializable {
+  Tracing.init()
```
Comment (ianoc) [inline scalding-core/src/main/scala/com/twitter/scalding/Job.scala]: Doesn't seem like it should be in the default job, something can be opted into by extending a trait instead

File: project/Build.scala [hunk 2]
```diff
@@ -237,7 +237,7 @@ object ScaldingBuild extends Build {
   lazy val scaldingDate = module("date")
 
   lazy val cascadingVersion =
-    System.getenv.asScala.getOrElse("SCALDING_CASCADING_VERSION", "2.5.5")
+    System.getenv.asScala.getOrElse("SCALDING_CASCADING_VERSION", "2.6.1")
 
   lazy val cascadingJDBCVersion =
     System.getenv.asScala.getOrElse("SCALDING_CASCADING_JDBC_VERSION", "2.5.4")
```
Comment (ianoc) [inline project/Build.scala]: Does 2.5 JDBC go with 2.6 cascading?

**Comment by JamesIry** [comment]
Understood. Let me know if there's anything I can do to help.

File: project/Build.scala (hunk 2 above)
Comment (JamesIry) [inline project/Build.scala]: It can, but it's not necessary.

File: scalding-core/src/main/scala/com/twitter/scalding/Job.scala (hunk 1 above)
Comment (JamesIry) [inline scalding-core/src/main/scala/com/twitter/scalding/Job.scala]: Are you suggesting that Job extend a trait that initializes Tracing? Or that end users should explicitly request tracing. If the later I disagree. This functionality isn't something people are likely to discover.

File: scalding-core/src/main/scala/com/twitter/scalding/Tracing.scala [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+object Tracing {
+  private val LOG: Logger = LogManager.getLogger(this.getClass)
+
+  /**
+   * Put a barrier at com.twitter.scalding, but exclude things like Tool
+   * that are common entry points for calling user code
+   */
+  private val defaultRegex = """^com\.twitter\.scalding\.(?!Tool|Job|ExecutionContext).*"""
+
+  register()
```
Comment (ianoc) [inline scalding-core/src/main/scala/com/twitter/scalding/Tracing.scala]: Why is this here and not in init?

File: scalding-core/src/main/scala/com/twitter/scalding/Job.scala (hunk 1 above)
Comment (ianoc) [inline scalding-core/src/main/scala/com/twitter/scalding/Job.scala]: The former, they should extend a trait. If they use driven they will know its there. Why have tracing code running for everyone as this would have? 

(Indeed it seems like tracing here should be a trait anyway that gets extended from. also can allow its methods easier overridden)

Comment (JamesIry) [inline scalding-core/src/main/scala/com/twitter/scalding/Job.scala]: A document service is a plugin deployed at runtime. The programmer may not have any idea that one is deployed.

The tracing code change is just registering a regex representing the boundary of the internals of Scalding. It runs once at start up and has no impact on users.

File: scalding-core/src/main/scala/com/twitter/scalding/Tracing.scala (hunk 3 above)
Comment (JamesIry) [inline scalding-core/src/main/scala/com/twitter/scalding/Tracing.scala]: It's a very cheap way to create one-time-init semantics.

**Comment by johnynek** [comment]
Okay, this seems safe enough, and good for Driven users.
